src/drivers/ammo-driver.js

```diff
@@ -1,10 +1,174 @@
+const Driver = require("./driver");
+window.AmmoModule = window.Ammo;
+window.Ammo = null;
 
-function AmmoDriver () {
-  throw new Error('[AmmoDriver] Driver not implemented.');
+const EPS = 10e-6;
+
+function AmmoDriver() {
+  this.collisionConfiguration = null;
+  this.dispatcher = null;
+  this.broadphase = null;
+  this.solver = null;
+  this.physicsWorld = null;
+  this.debugDrawer = null;
+
+  this.els = {};
+  this.eventListeners = [];
+  this.collisions = {};
+  this.currentCollisions = [];
 }
 
 AmmoDriver.prototype = new Driver();
 AmmoDriver.prototype.constructor = AmmoDriver;
 
 module.exports = AmmoDriver;
+
+/* @param {object} worldConfig */
+AmmoDriver.prototype.init = function(worldConfig) {
+  return new Promise(resolve => {
+    AmmoModule().then(result => {
+      Ammo = result;
+      this.epsilon = worldConfig.epsilon || EPS;
+      this.debugDrawMode = worldConfig.debugDrawMode || THREE.AmmoDebugConstants.NoDebug;
+      this.maxSubSteps = worldConfig.maxSubSteps || 4;
+      this.fixedTimeStep = worldConfig.fixedTimeStep || 1 / 60;
+      this.collisionConfiguration = new Ammo.btDefaultCollisionConfiguration();
+      this.dispatcher = new Ammo.btCollisionDispatcher(this.collisionConfiguration);
+      this.broadphase = new Ammo.btDbvtBroadphase();
+      this.solver = new Ammo.btSequentialImpulseConstraintSolver();
+      this.physicsWorld = new Ammo.btDiscreteDynamicsWorld(
+        this.dispatcher,
+        this.broadphase,
+        this.solver,
+        this.collisionConfiguration
+      );
+      this.physicsWorld.setGravity(
+        new Ammo.btVector3(0, worldConfig.hasOwnProperty("gravity") ? worldConfig.gravity : -9.8, 0)
+      );
+      this.physicsWorld.getSolverInfo().set_m_numIterations(worldConfig.solverIterations);
+      resolve();
+    });
+  });
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.addBody = function(body, group, mask) {
+  this.physicsWorld.addRigidBody(body, group, mask);
+  this.els[Ammo.getPointer(body)] = body.el;
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.removeBody = function(body) {
+  this.physicsWorld.removeRigidBody(body);
+  this.removeEventListener(body);
+  delete this.els[Ammo.getPointer(body)];
+};
+
+AmmoDriver.prototype.updateBody = function(body) {
+  if (this.els.hasOwnProperty(Ammo.getPointer(body))) {
+    this.physicsWorld.updateSingleAabb(body);
+  }
+};
+
+/* @param {number} deltaTime */
+AmmoDriver.prototype.step = function(deltaTime) {
+  this.physicsWorld.stepSimulation(deltaTime, this.maxSubSteps, this.fixedTimeStep);
+
+  const numManifolds = this.dispatcher.getNumManifolds();
+  for (let i = 0; i < numManifolds; i++) {
+    const persistentManifold = this.dispatcher.getManifoldByIndexInternal(i);
+    const numContacts = persistentManifold.getNumContacts();
+    const body0ptr = Ammo.getPointer(persistentManifold.getBody0());
+    const body1ptr = Ammo.getPointer(persistentManifold.getBody1());
```

Comment:
FYI, in Emscripten, if you ask for an Emscripten wrapper object like this, Emscripten keeps a big cache mapping `{ptr: wrapper}` and will reuse an existing wrapper iff it's seen the pointer you want a wrapper for before.

So for the body objects returned by `getBody0` and `getBody1` here, since those are clearly going to be the same for the lifetime of the rigid body, those shouldn't allocate new wrapper objects every time. But for the manifold object returned by `getManifoldByIndexInternal`, for example, or the manifold point returned by `getContactPoint`, if those pointers are new for each physics step, it will allocate a new wrapper object every time, which probably isn't what you want.

If that's the case, and you care enough about this stuff, you could fix it in performance-sensitive spots like this by manipulating the heap directly instead of using the wrapper objects. You could also fix it by calling the C method directly and using your own object pooling for the wrapper, e.g. something like:

```javascript
// this.manifold is already an `Ammo.btPersistentManifold` that I want to reuse
this.manifold.ptr = Ammo._emscripten_bind_btDispatcher_getManifoldByIndexInternal_1(self, this.dispatcher.ptr)
```

Reply:
I'm not sure if this exactly resolves your concern, but I did see that the ptrs used by persistentManifolds get reused, which is why I didn't just key off of those pointers instead of a combination of the body ptrs. If that's the case, then the wrappers should be getting reused too?

Reply:
Great, yeah, if Ammo reuses the pointers then Emscripten will reuse the wrapper objects.